src/bio/bio_recovery.c

```diff
@@ -197,54 +222,81 @@ on_teardown(struct bio_blobstore *bbs)
 }
 
 static void
-setup_xstream(void *arg)
+setup_blobstore(struct bio_xs_context *xs_ctxt, enum smd_type st,
+		int *closed_blobs)
 {
-	struct bio_xs_context	*xs_ctxt = arg;
-	struct bio_blobstore	*bbs;
 	struct bio_io_context	*ioc;
-	int			 closed_blobs = 0;
+	struct bio_blobstore	*bbs;
+	struct bio_xs_blobstore	*bxb;
 
-	D_ASSERT(xs_ctxt != NULL);
-	if (!is_server_started()) {
-		D_INFO("Abort xs setup on server start/shutdown\n");
+	bxb = xs_ctxt->bxc_xs_blobstores[st];
+	if (bxb == NULL)
 		return;
-	}
 
-	bbs = xs_ctxt->bxc_blobstore;
-	D_ASSERT(bbs != NULL);
-	D_ASSERT(bbs->bb_bs != NULL);
+	bbs = bxb->bxb_blobstore;
+	if (bbs == NULL)
+		return;
 
+	D_ASSERT(bbs->bb_bs != NULL);
 	/*
 	 * Setup the blobstore io channel. It's must be done as the first step
-	 * of xstream setup, since xstream teardown checks io channel to tell
-	 * if everything is torndown for the xstream.
+	 * of xstream setup, since blobstore teardown checks io channel to tell
+	 * if everything is torndown for the blobstore.
 	 */
-	if (xs_ctxt->bxc_io_channel == NULL) {
-		xs_ctxt->bxc_io_channel = spdk_bs_alloc_io_channel(bbs->bb_bs);
-		if (xs_ctxt->bxc_io_channel == NULL) {
+	if (bxb->bxb_io_channel == NULL) {
+		bxb->bxb_io_channel = spdk_bs_alloc_io_channel(bbs->bb_bs);
+		if (bxb->bxb_io_channel == NULL) {
 			D_ERROR("Failed to create io channel for %p\n", bbs);
 			return;
 		}
 	}
 
 	/* Try to open all blobs */
-	d_list_for_each_entry(ioc, &xs_ctxt->bxc_io_ctxts, bic_link) {
+	d_list_for_each_entry(ioc, &bxb->bxb_io_ctxts, bic_link) {
 		if (ioc->bic_blob != NULL && !ioc->bic_closing)
 			continue;
 
-		closed_blobs++;
+		*closed_blobs = *closed_blobs + 1;
 		if (ioc->bic_opening || ioc->bic_closing)
 			continue;
 
 		bio_blob_open(ioc, true);
 	}
 
-	if (closed_blobs) {
-		D_DEBUG(DB_MGMT, "xs:%p has %d closed blobs\n",
-			xs_ctxt, closed_blobs);
+	if (*closed_blobs)
+		D_DEBUG(DB_MGMT, "blobstore:%p has %d closed blobs\n",
+			bbs, *closed_blobs);
+	return;
+
+}
+
+static void
+setup_xstream(void *arg)
```

Comment:
I think we can support NVMe recovery for meta and WAL SSDs later, for this moment, we only do recovery for data SSD, what do you think?

Reply:
sure, makes sense.